watertap/unit_models/gac.py

```diff
@@ -760,209 +708,303 @@ def eq_dg(b, t, j):
+        def eq_number_bi(b):
             return b.N_Bi * b.ds * b.dg * b.bed_voidage == b.kf * (
                 b.particle_dia / 2
             ) * (1 - b.bed_voidage)
 
+        # ---------------------------------------------------------------------
+        # bed dimensions, gac particle, and sizing calculations
+
+        @self.Constraint(doc="bed void fraction based on gac particle densities")
+        def eq_bed_voidage(b):
+            return b.bed_voidage == 1 - (b.particle_dens_bulk / b.particle_dens_app)
+
+        @self.Constraint(doc="relating velocities based on bed voidage")
+        def eq_velocity_relation(b):
+            return b.velocity_sup == b.velocity_int * b.bed_voidage
+
+        @self.Constraint(doc="bed length based on velocity and ebct")
+        def eq_bed_length(b):
+            return b.bed_length == b.velocity_sup * b.ebct
+
+        @self.Constraint(doc="bed diameter and area relation")
+        def eq_bed_diameter(b):
+            return Constants.pi * (b.bed_diameter**2) == 4 * b.bed_area
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="bed area based on velocity and volumetric flow",
+        )
+        def eq_bed_area(b, t):
+            return (
+                b.bed_area * b.velocity_sup
+                == b.process_flow.properties_in[t].flow_vol_phase["Liq"]
+            )
+
+        @self.Constraint(doc="bed volume based on cylindrical dimensions")
+        def eq_bed_volume(b):
+            return b.bed_volume == b.bed_length * b.bed_area
+
+        @self.Constraint(doc="fluid residence time in the bed")
+        def eq_residence_time(b):
+            return b.residence_time == b.bed_voidage * b.ebct
+
+        @self.Constraint(doc="total mass of gac in the bed")
+        def eq_mass_gac_bed(b):
+            return b.bed_mass_gac == b.particle_dens_bulk * b.bed_volume
+
+        # ---------------------------------------------------------------------
+        # gac CPHSDM intermediate equations
+
         @self.Constraint(
-            doc="Minimum Stanton number to achieve constant pattern solution"
+            doc="minimum Stanton number to achieve constant pattern solution"
         )
-        def eq_min_n_st_cps(b):
+        def eq_min_number_st_cps(b):
             return b.min_N_St == b.a0 * b.N_Bi + b.a1
 
         @self.Constraint(
-            doc="Minimum empty bed contact time to achieve constant pattern solution"
+            doc="minimum empty bed contact time to achieve constant pattern solution"
         )
         def eq_min_ebct_cps(b):
             return b.min_ebct * (1 - b.bed_voidage) * b.kf == b.min_N_St * (
                 b.particle_dia / 2
             )
 
+        @self.Constraint(doc="throughput based on empirical 5-parameter regression")
+        def eq_throughput(b):
+            return b.throughput == b.b0 + b.b1 * (
+                b.conc_ratio_replace**b.b2
+            ) + b.b3 / (1.01 - (b.conc_ratio_replace**b.b4))
+
```

Comment:
Going through this quickly, but I see this constraint pop up twice, except one is applied to discretized elements and the other is this one, which probably works out. I haven't connected all the dots between constraints though. After doing this model overhaul, did you check for degeneracy again after solving successfully?

Reply:
Good news is I just double checked and DegeneracyHunter says it is in the clear, but there is a more efficient way to implement this, though slightly challenging.

So in the discretization of this equation, this is basically the end point of the steps. So this solution is used to determine the step size given a fixed number of discretized elements. I played around with changing the indices (which are already a little convoluted) and declaring a rule that fixing the matching element_[n] case to the original case, but it turned out to be more difficult than expected. If I remember correctly, it would be solving the performance equations for elements i throguh (n-1) but solve the linspace utility for element n because of how the variables would be fixed. I think it's worth revisiting in subsequent PR's just because the amount of changes in this. Let me know your thoughts.